File: db-guard/node/src/contract.ts
import type { PrismaDbGuardOptions } from './prisma';
import type { MongooseDbGuardOptions } from './mongoose';
import type { DrizzleDbGuardOptions } from './drizzle';
import type { TypeOrmDbGuardOptions } from './typeorm';
import type { RateLimiterOptions } from './security';

export const DB_GUARD_CONTRACT_VERSION = 1 as const;
export const SUPPORTED_KEY_VERSIONS = ['1', '2'] as const;

export type DbGuardContractErrorCode =
  | 'UNSUPPORTED_CONTRACT_VERSION'
  | 'INVALID_KEYRING'
  | 'INVALID_RESOURCE_SCOPE'
  | 'RESOURCE_NOT_FOUND';

export class DbGuardContractError extends Error {
  constructor(
    public readonly code: DbGuardContractErrorCode,
    message: string
  ) {
    super(`Vollcrypt DbGuard contract [${code}]: ${message}`);
    this.name = 'DbGuardContractError';
  }
}

export type DbGuardMask =
  | 'credit_card'
  | 'email'
  | 'tc_no'
  | ((value: any) => any)
  | string;

export interface CryptoRbacConfig {
  roles: Record<string, {
    decrypt: string[];
    mask?: Record<string, DbGuardMask>;
  }>;
}

export interface CommonDbGuardSecurityOptions {
  cryptoRbac?: CryptoRbacConfig;
  rateLimiter?: RateLimiterOptions;
  allowUnrestrictedDecrypt?: boolean;
}

export interface DbGuardContractV1 {
  contractVersion: typeof DB_GUARD_CONTRACT_VERSION;
  keyring: {
    keys: Buffer | Record<string, Buffer>;
    activeVersion?: string;
  };
  resources: Record<string, string[]>;
  blindIndexes?: {
    rootSalt: Buffer;
    allowFrequencyLeakage: true;
    resources: Record<string, string[]>;
  };
  security?: CommonDbGuardSecurityOptions;
}

function keyVersions(keys: Buffer | Record<string, Buffer>): string[] {
  return Buffer.isBuffer(keys) ? ['1'] : Object.keys(keys);
}

function validateResourceMap(resources: Record<string, string[]>, label: string) {
  const entries = Object.entries(resources);
  if (entries.length === 0) {
    throw new DbGuardContractError('INVALID_RESOURCE_SCOPE', `${label} must not be empty.`);
  }
  for (const [resource, fields] of entries) {
    if (!resource.trim() || !Array.isArray(fields) || fields.length === 0) {
      throw new DbGuardContractError(
        'INVALID_RESOURCE_SCOPE',
        `${label}.${resource || '<empty>'} must contain at least one field.`
      );
    }
    if (fields.some((field) => typeof field !== 'string' || !field.trim())) {
      throw new DbGuardContractError(
        'INVALID_RESOURCE_SCOPE',
        `${label}.${resource} contains an invalid field name.`
      );
    }
  }
}

export function validateDbGuardContract(contract: DbGuardContractV1): void {
  if (contract.contractVersion !== DB_GUARD_CONTRACT_VERSION) {
    throw new DbGuardContractError(
      'UNSUPPORTED_CONTRACT_VERSION',
      `expected ${DB_GUARD_CONTRACT_VERSION}, got ${String(contract.contractVersion)}.`
    );
  }

  const versions = keyVersions(contract.keyring.keys);
  if (versions.length === 0 || versions.some((version) => !SUPPORTED_KEY_VERSIONS.includes(version as any))) {
    throw new DbGuardContractError(
      'INVALID_KEYRING',
      `key versions must be one or more of ${SUPPORTED_KEY_VERSIONS.join(', ')}.`
    );
  }
  for (const version of versions) {
    const key = Buffer.isBuffer(contract.keyring.keys)
      ? contract.keyring.keys
      : contract.keyring.keys[version];
    if (!Buffer.isBuffer(key) || key.length !== 32) {
      throw new DbGuardContractError('INVALID_KEYRING', `key version ${version} must be 32 bytes.`);
    }
  }

  const activeVersion = contract.keyring.activeVersion || versions[0];
  if (!versions.includes(activeVersion)) {
    throw new DbGuardContractError(
      'INVALID_KEYRING',
      `active key version ${activeVersion} is not present in the keyring.`
    );
  }

  validateResourceMap(contract.resources, 'resources');
  if (contract.blindIndexes) {
    validateResourceMap(contract.blindIndexes.resources, 'blindIndexes.resources');
  }
}

function securityOptions(contract: DbGuardContractV1): CommonDbGuardSecurityOptions {
  return contract.security ? { ...contract.security } : {};
}

export function toPrismaDbGuardOptions(contract: DbGuardContractV1): PrismaDbGuardOptions {
  validateDbGuardContract(contract);
  return {
    key: contract.keyring.keys,
    activeKeyVersion: contract.keyring.activeVersion,
    models: contract.resources,
    blindIndexes: contract.blindIndexes
      ? {
          rootSalt: contract.blindIndexes.rootSalt,
          allowFrequencyLeakage: contract.blindIndexes.allowFrequencyLeakage,
          models: contract.blindIndexes.resources
        }
      : undefined,
    ...securityOptions(contract)
  };
}

export function toMongooseDbGuardOptions(
  contract: DbGuardContractV1,
  resource: string
): MongooseDbGuardOptions {
  validateDbGuardContract(contract);
  const fields = contract.resources[resource];
  if (!fields) {
    throw new DbGuardContractError('RESOURCE_NOT_FOUND', `resource ${resource} is not configured.`);
  }
  return {
    key: contract.keyring.keys,
    activeKeyVersion: contract.keyring.activeVersion,
    fields,
    blindIndexes: contract.blindIndexes
      ? {
          rootSalt: contract.blindIndexes.rootSalt,
          allowFrequencyLeakage: contract.blindIndexes.allowFrequencyLeakage,
          fields: contract.blindIndexes.resources[resource] || [],
          modelName: resource
        }
      : undefined,
    ...securityOptions(contract)
  };
}

export function toDrizzleDbGuardOptions(contract: DbGuardContractV1): DrizzleDbGuardOptions {
  validateDbGuardContract(contract);
  return {
    key: contract.keyring.keys,
    activeKeyVersion: contract.keyring.activeVersion,
    blindIndexes: contract.blindIndexes
      ? {
          rootSalt: contract.blindIndexes.rootSalt,
          allowFrequencyLeakage: contract.blindIndexes.allowFrequencyLeakage
        }
      : undefined,
    ...securityOptions(contract)
  };
}

export function toTypeOrmDbGuardOptions(contract: DbGuardContractV1): TypeOrmDbGuardOptions {
  validateDbGuardContract(contract);
  return {
    key: contract.keyring.keys,
    activeKeyVersion: contract.keyring.activeVersion,
    entities: contract.resources,
    blindIndexes: contract.blindIndexes
      ? {
          rootSalt: contract.blindIndexes.rootSalt,
          allowFrequencyLeakage: contract.blindIndexes.allowFrequencyLeakage,
          entities: contract.blindIndexes.resources
        }
      : undefined,
    ...securityOptions(contract)
  };
}

File: db-guard/rust/src/contract.rs
use std::collections::{BTreeMap, HashMap};
use std::error::Error;
use std::fmt;

pub const DB_GUARD_CONTRACT_VERSION: u8 = 1;
pub const SUPPORTED_KEY_VERSIONS: [&str; 2] = ["1", "2"];

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DbGuardContractV1 {
    pub contract_version: u8,
    pub active_key_version: String,
    pub resources: BTreeMap<String, Vec<String>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DbGuardContractError {
    pub code: &'static str,
    message: String,
}

impl DbGuardContractError {
    fn new(code: &'static str, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

impl fmt::Display for DbGuardContractError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Vollcrypt DbGuard contract [{}]: {}",
            self.code, self.message
        )
    }
}

impl Error for DbGuardContractError {}

pub fn validate_orm_contract(
    contract: &DbGuardContractV1,
    keys: &HashMap<String, Vec<u8>>,
) -> Result<(), DbGuardContractError> {
    if contract.contract_version != DB_GUARD_CONTRACT_VERSION {
        return Err(DbGuardContractError::new(
            "UNSUPPORTED_CONTRACT_VERSION",
            format!(
                "expected {}, got {}.",
                DB_GUARD_CONTRACT_VERSION, contract.contract_version
            ),
        ));
    }

    if keys.is_empty() {
        return Err(DbGuardContractError::new(
            "INVALID_KEYRING",
            "keyring must not be empty.",
        ));
    }
    for (version, key) in keys {
        if !SUPPORTED_KEY_VERSIONS.contains(&version.as_str()) {
            return Err(DbGuardContractError::new(
                "INVALID_KEYRING",
                format!("unsupported key version {version}."),
            ));
        }
        if key.len() != 32 {
            return Err(DbGuardContractError::new(
                "INVALID_KEYRING",
                format!("key version {version} must be 32 bytes."),
            ));
        }
    }
    if !keys.contains_key(&contract.active_key_version) {
        return Err(DbGuardContractError::new(
            "INVALID_KEYRING",
            format!(
                "active key version {} is not present in the keyring.",
                contract.active_key_version
            ),
        ));
    }

    if contract.resources.is_empty() {
        return Err(DbGuardContractError::new(
            "INVALID_RESOURCE_SCOPE",
            "resources must not be empty.",
        ));
    }
    for (resource, fields) in &contract.resources {
        if resource.trim().is_empty()
            || fields.is_empty()
            || fields.iter().any(|field| field.trim().is_empty())
        {
            return Err(DbGuardContractError::new(
                "INVALID_RESOURCE_SCOPE",
                format!("resource {resource:?} must contain valid field names."),
            ));
        }
    }

    Ok(())
}

pub fn configure_orm_contract(
    contract: &DbGuardContractV1,
    keys: &HashMap<String, Vec<u8>>,
) -> Result<(), DbGuardContractError> {
    validate_orm_contract(contract, keys)?;

    crate::clear_registry();
    for (version, key) in keys {
        crate::set_key(version, key);
    }
    crate::set_active_version(&contract.active_key_version)
        .map_err(|message| DbGuardContractError::new("INVALID_KEYRING", message))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn contract() -> DbGuardContractV1 {
        DbGuardContractV1 {
            contract_version: DB_GUARD_CONTRACT_VERSION,
            active_key_version: "2".to_string(),
            resources: BTreeMap::from([(
                "User".to_string(),
                vec!["email".to_string(), "card".to_string()],
            )]),
        }
    }

    fn keys() -> HashMap<String, Vec<u8>> {
        HashMap::from([
            ("1".to_string(), vec![1; 32]),
            ("2".to_string(), vec![2; 32]),
        ])
    }

    #[test]
    fn validates_the_shared_diesel_and_seaorm_contract() {
        assert_eq!(validate_orm_contract(&contract(), &keys()), Ok(()));
    }

    #[test]
    fn rejects_contract_version_keyring_and_resource_mismatches() {
        let mut invalid_version = contract();
        invalid_version.contract_version = 2;
        assert_eq!(
            validate_orm_contract(&invalid_version, &keys())
                .expect_err("version must fail")
                .code,
            "UNSUPPORTED_CONTRACT_VERSION"
        );

        let mut missing_active = keys();
        missing_active.remove("2");
        assert_eq!(
            validate_orm_contract(&contract(), &missing_active)
                .expect_err("missing active key must fail")
                .code,
            "INVALID_KEYRING"
        );

        let mut empty_resources = contract();
        empty_resources.resources.clear();
        assert_eq!(
            validate_orm_contract(&empty_resources, &keys())
                .expect_err("empty resources must fail")
                .code,
            "INVALID_RESOURCE_SCOPE"
        );
    }
}
